From c9ad1822fd34b06cb9703abb01ab0b3e7f28cfee Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Fri, 22 Apr 2022 15:40:59 +0200
Subject: [PATCH 165/841] feat(stm32mp2): syscfg: Add IO compensation handling

Add a function to enable IO compensation for an IO domain.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 2597d12b08ed6d87e5dd7d6e517d0d5ae6dccdf8 feat(stm32mp2): syscfg: Add IO compensation handling]
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Change-Id: I7dc0f0fc31c259c85aafcbe25c79d6f8e57eced7
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/249113
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp2/include/stm32mp2_private.h | 12 ++++
 plat/st/stm32mp2/stm32mp2_syscfg.c          | 65 +++++++++++++++++++++
 2 files changed, 77 insertions(+)

diff --git a/plat/st/stm32mp2/include/stm32mp2_private.h b/plat/st/stm32mp2/include/stm32mp2_private.h
index cdbc3ef99..ef500f067 100644
--- a/plat/st/stm32mp2/include/stm32mp2_private.h
+++ b/plat/st/stm32mp2/include/stm32mp2_private.h
@@ -16,6 +16,18 @@ uint32_t stm32mp2_syscfg_get_chip_version(void);
 uint32_t stm32mp2_syscfg_get_chip_dev_id(void);
 size_t stm32mp2_syscfg_get_mm_size(uint8_t bank);
 
+/* IO compensation identifiers */
+enum syscfg_io_ids {
+	SYSFG_VDDIO1_ID,
+	SYSFG_VDDIO2_ID,
+	SYSFG_VDDIO3_ID,
+	SYSFG_VDDIO4_ID,
+	SYSFG_VDD_IO_ID,
+	SYSFG_NB_IO_ID
+};
+
+void stm32mp2_syscfg_enable_io_compensation(enum syscfg_io_ids id);
+
 /* Get RISAF platform instance ID from peripheral IO memory base address */
 int stm32_risaf_get_instance(uintptr_t base);
 
diff --git a/plat/st/stm32mp2/stm32mp2_syscfg.c b/plat/st/stm32mp2/stm32mp2_syscfg.c
index 88ef36cce..ef2205198 100644
--- a/plat/st/stm32mp2/stm32mp2_syscfg.c
+++ b/plat/st/stm32mp2/stm32mp2_syscfg.c
@@ -7,6 +7,7 @@
 #include <platform_def.h>
 
 #include <common/debug.h>
+#include <drivers/delay_timer.h>
 #include <lib/mmio.h>
 #include <lib/utils_def.h>
 
@@ -30,6 +31,39 @@
 #define SYSCFG_DEVICEID_REV_ID_MASK	GENMASK_32(31, 16)
 #define SYSCFG_DEVICEID_REV_ID_SHIFT	16
 
+/*
+ * SYSCFG IO Compensation Registers
+ */
+#define SYSCFG_VDDIO3CCCR		0x4000U
+#define SYSCFG_VDDIO4CCCR		0x4008U
+#define SYSCFG_VDDCCCR			0x4010U
+#define SYSCFG_VDDIO2CCCR		0x4018U
+#define SYSCFG_VDDIO1CCCR		0x4020U
+
+/* IO compensation CCR registers bit definition */
+#define SYSCFG_CCCR_CS			BIT(9)
+#define SYSCFG_CCCR_EN			BIT(8)
+#define SYSCFG_CCCR_RAPSRC_MASK		GENMASK(7, 4)
+#define SYSCFG_CCCR_RANSRC_MASK		GENMASK(3, 0)
+
+/* IO compensation CCSR registers bit definition */
+#define SYSCFG_CCSR_READY		BIT(8)
+#define SYSCFG_CCSR_APSRC_MASK		GENMASK(7, 4)
+#define SYSCFG_CCSR_ANSRC_MASK		GENMASK(3, 0)
+
+#define SYSCFG_CCSR_READY_TIMEOUT_US	1000U
+
+/*
+ * SYSCFG IO compensation register offsets (base relative)
+ */
+static uint32_t syscfg_cccr_offset[SYSFG_NB_IO_ID] = {
+	[SYSFG_VDDIO1_ID] = SYSCFG_VDDIO1CCCR,
+	[SYSFG_VDDIO2_ID] = SYSCFG_VDDIO2CCCR,
+	[SYSFG_VDDIO3_ID] = SYSCFG_VDDIO3CCCR,
+	[SYSFG_VDDIO4_ID] = SYSCFG_VDDIO4CCCR,
+	[SYSFG_VDD_IO_ID] = SYSCFG_VDDCCCR,
+};
+
 /*
  * @brief  Get silicon revision from SYSCFG registers.
  * @retval chip version (REV_ID).
@@ -78,3 +112,34 @@ size_t stm32mp2_syscfg_get_mm_size(uint8_t bank)
 
 	return (bank == 0U) ? addr_mapping1 : addr_mapping2;
 }
+
+/*
+ * @brief  Enable IO compensation for an IO domain.
+ * @param  id: IO compensation ID
+ * @retval 0.
+ */
+void stm32mp2_syscfg_enable_io_compensation(enum syscfg_io_ids id)
+{
+	uintptr_t cccr_addr = SYSCFG_BASE + syscfg_cccr_offset[id];
+	uintptr_t ccsr_addr = cccr_addr + 4U;
+	uint64_t timeout_ref;
+
+	VERBOSE("Enable IO comp for id %u\n", id);
+
+	if ((mmio_read_32(ccsr_addr) & SYSCFG_CCSR_READY) != 0U) {
+		return;
+	}
+
+	mmio_setbits_32(cccr_addr, SYSCFG_CCCR_EN);
+
+	timeout_ref = timeout_init_us(SYSCFG_CCSR_READY_TIMEOUT_US);
+
+	while ((mmio_read_32(ccsr_addr) & SYSCFG_CCSR_READY) == 0U)
+		if (timeout_elapsed(timeout_ref)) {
+			WARN("IO compensation cell not ready\n");
+			/* Allow an almost silent failure here */
+			break;
+		}
+
+	mmio_clrbits_32(cccr_addr, SYSCFG_CCCR_CS);
+}
-- 
2.47.2

